### PR TITLE
feat: add typescript declaration file

### PR DESCRIPTION
**Pull request by SteveBunlon**
## Pull Request checklist:

- [x] Write an explicit title for the Pull Request, following [Conventional Commits specification](https://www.conventionalcommits.org)
- [x] Create automatic tests
- [ ] Test manually the implemented changes
- [x] Review my own code (indentation, syntax, style, simplicity, readability)
- [x] Wonder if you can improve the existing code

## Disclaimer

The is the first attempt of adding Types definition in our repo directly. 

### Observations

The `how to translate your app into typescript` page on the Woodshop is, by far, the second most consulted page. 
Some of our top paying clients using mongoose develops using Typescript

### Problems encountered

We recently have been addressed some issues regarding our Types definition in the community (see this [card](https://app.clickup.com/t/kn5vb3) for more info)

Maintaining Types in DefinitelyTyped has been done using a _community_ approach till now. This leads to some frustration when users upgrade their lianas without updates on their types. Some of our dependencies also have stoped working DefinitelyTyped and integrate their own Types directly in their repos, which makes hard if not impossible to synchronise our own Types. Maintaining Types DefinitelyTyped also leaded to some frustration from us, because updating Types can take up to three weeks to be validated by maintainers (they have a lot to do !). 

## About this version

As stated before, this is the very attempt to add Types in our own repos. Every comments on structure, declarations, scaffolding and naming are welcomed.

## How to test

You can use the following project to test the types: https://github.com/SteveBunlon/fa-blog-ts-postgres
Follow the `README.md` to init the project, then try to use the type (eg. create smart actions, override routes...) and check that auto-completion works as intended.

### CODE REVIEW

File: types/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+  description?: string;
+  type: string | string[];
+  isReadOnly?: boolean;
+  isRequired?: boolean;
+  reference?: string;
+  enums?: string[];
+  defaultValue?: any;
+  get?: SmartFieldValueGetter;
+  set?: SmartFieldValueSetter;
+  search?: SmartFieldSearcher;
```
Comment (olesyakorovina) [inline types/index.d.ts:188]: `filter` property is missing

**Review by olesyakorovina — CHANGES_REQUESTED** [review]

File: types/index.d.ts (hunk 1 above)
Comment (SteveBunlon) [reply types/index.d.ts:188]: 👍 🙏  Added

**Review by olesyakorovina — APPROVED** [review]
🔥

File: types/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+}
+
+export interface SmartFieldOptions {
+  field: string;
+  description?: string;
+  type: string | string[];
+  isReadOnly?: boolean;
+  isRequired?: boolean;
+  reference?: string;
+  enums?: string[];
```
[the remark this answers is not in the document]
Comment (SteveBunlon) [reply types/index.d.ts]: 👍

File: types/index.d.ts [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+}
+
+export interface SmartActionField {
+  field: string,
+  description?: string,
+  type: FieldType,
+  isRequired?: boolean,
+  isReadOnly?: boolean,
+  enums?: FieldEnumsType,
+  defaultValue?: any,
```
[the remark this answers is not in the document]
Comment (SteveBunlon) [reply types/index.d.ts:204]: 👍

File: types/index.d.ts [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+
+type FieldType = 'Boolean' | 'Date' | 'Dateonly' | 'Enum' | 'File' | 'Number' | 'String' | ['Enum'] | ['Number'] | ['String'];
+
+type FieldEnumsType = string[] | number[] | Date[] | boolean[];
+
+export interface SmartFieldOptions {
+  field: string;
+  description?: string;
+  type: FieldType;
+  isFilterable: boolean;
```
[the remark this answers is not in the document]
Comment (SteveBunlon) [reply types/index.d.ts:185]: 👍

**Review by olesyakorovina — APPROVED** [review]

**Review by olesyakorovina — APPROVED** [review]

**Pull request merged**